src/backend/distributed/executor/merge_executor.c

```diff
@@ -0,0 +1,422 @@
+ * data source.
+ */
+void
+ExecuteSourceAtCoordAndRedistribution(CitusScanState *scanState)
+{
+	EState *executorState = ScanStateGetExecutorState(scanState);
+	DistributedPlan *distributedPlan = scanState->distributedPlan;
+	Query *mergeQuery =
+		copyObject(distributedPlan->modifyQueryViaCoordinatorOrRepartition);
+	RangeTblEntry *targetRte = ExtractResultRelationRTE(mergeQuery);
+	RangeTblEntry *sourceRte = ExtractMergeSourceRangeTableEntry(mergeQuery);
+	Query *sourceQuery = sourceRte->subquery;
+	Oid targetRelationId = targetRte->relid;
+	PlannedStmt *sourcePlan =
+		copyObject(distributedPlan->selectPlanForModifyViaCoordinatorOrRepartition);
+	char *intermediateResultIdPrefix = distributedPlan->intermediateResultIdPrefix;
+	bool hasReturning = distributedPlan->expectResults;
+
+	/*
+	 * If we are dealing with partitioned table, we also need to lock its
+	 * partitions. Here we only lock targetRelation, we acquire necessary
+	 * locks on source tables during execution of those source queries.
+	 */
+	if (PartitionedTable(targetRelationId))
+	{
+		LockPartitionRelations(targetRelationId, RowExclusiveLock);
+	}
+
+	ereport(DEBUG1, (errmsg("Collect source query results on coordinator")));
+
+	List *prunedTaskList = NIL;
+	HTAB *shardStateHash =
+		ExecuteMergeSourcePlanIntoColocatedIntermediateResults(
+			targetRelationId,
+			mergeQuery,
+			sourceQuery->targetList,     /*TODO */
+			sourcePlan,
+			executorState,
+			intermediateResultIdPrefix);
+
+	ereport(DEBUG1, (errmsg("Create a MERGE task list that needs to be routed")));
+
+	/* generate tasks for the .. phase */
+	List *taskList =
+		GenerateTaskListWithColocatedIntermediateResults(targetRelationId, mergeQuery,
+														 intermediateResultIdPrefix);
+
+	/*
+	 * We cannot actually execute MERGE INTO ... tasks that read from
+	 * intermediate results that weren't created because no rows were
+	 * written to them. Prune those tasks out by only including tasks
+	 * on shards with connections.
+	 */
+	Task *task = NULL;
+	foreach_ptr(task, taskList)
+	{
+		uint64 shardId = task->anchorShardId;
+		bool shardModified = false;
+
+		hash_search(shardStateHash, &shardId, HASH_FIND, &shardModified);
+		if (shardModified)
+		{
+			prunedTaskList = lappend(prunedTaskList, task);
+		}
+	}
+
+	if (prunedTaskList == NIL)
+	{
+		/* No task to execute */
+		return;
+	}
+
+	ereport(DEBUG1, (errmsg("Execute MERGE task list")));
+	bool randomAccess = true;
+	bool interTransactions = false;
+	Assert(scanState->tuplestorestate == NULL);
+	scanState->tuplestorestate = tuplestore_begin_heap(randomAccess, interTransactions,
+													   work_mem);
+	TupleDesc tupleDescriptor = ScanStateGetTupleDescriptor(scanState);
+	TupleDestination *tupleDest =
+		CreateTupleStoreTupleDest(scanState->tuplestorestate, tupleDescriptor);
+	uint64 rowsMerged =
+		ExecuteTaskListIntoTupleDest(ROW_MODIFY_COMMUTATIVE, prunedTaskList,
```

Comment:
I believe we should mark `ROW_MODIFY_COMMUTATIVE` because it can do UPDATE/DELETEs.  Same for the other case as well

Well, we already prevent concurrent MERGE commands on the same target in AcquireExecutorShardLocksForExecution but still let's properly mark these

Reply:
@onderkalaci  I'm confused, you meant `ROW_MODIFY_NONCOMMUTATIVE `?

Reply:
Done.